### PR TITLE
Implement remaining missing operations in native ONNX parser

### PR DESCRIPTION
**Pull request by MatthewDaggitt**
Also fixes a couple of bugs in the parser I found and adds a better error message for when the user forgets to provide a network file on the command line.

### CODE REVIEW

File: src/engine/Marabou.cpp [hunk 1]
```diff
@@ -93,6 +93,12 @@ void Marabou::prepareInputQuery()
         */
         String networkFilePath = Options::get()->getString( Options::INPUT_FILE_PATH );
 
+        if ( networkFilePath.length() == 0 )
+        {
+            printf( "Error: no network file provided!\n" );
+            throw MarabouError( MarabouError::FILE_DOESNT_EXIST, networkFilePath.ascii() );
+        }
+
         if ( !File::exists( networkFilePath ) )
```
Comment (anwu1219) [inline src/engine/Marabou.cpp:102]: Change this line to "else if" then?

**Review by anwu1219 — CHANGES_REQUESTED** [review]
Hi Matthew, thanks for this contribution. 
The implementation looks correct to me. However, given the amount of details in the conv operation, is it possible to add some unit tests to the onnx parser?
One way might be to load a small convolutional network, set the input to constant, solve it and see if the output variable assignment matches the expectation.

Otherwise, the PR looks good to me.

Here are some unit tests that we have for the onnx parser implemented in the python API. I think those networks are small enough to be considered for the unit test.

https://github.com/NeuralNetworkVerification/Marabou/blob/663835df7b56be42af94b3d2723bf1b69494c0d5/maraboupy/test/test_onnx.py#L67-L74

https://github.com/NeuralNetworkVerification/Marabou/blob/663835df7b56be42af94b3d2723bf1b69494c0d5/maraboupy/test/test_onnx.py#L50-L57

File: src/engine/Marabou.cpp (hunk 1 above)
Comment (MatthewDaggitt) [reply src/engine/Marabou.cpp:102]: I'd prefer not to. I think early return points are much cleaner to read and require less mental effort. However, if you insist I can change it?

**Comment by MatthewDaggitt** [comment]
Thanks for the review!

> However, given the amount of details in the conv operation, is it possible to add some unit tests to the onnx parser?

Yup sure! I'm trying to add your example networks now, but I'm having to implement some of the missing operators such as `transpose` as well.

**Comment by MatthewDaggitt** [comment]
Added the missing operation implementations. Will add the tests later this week.

**Comment by MatthewDaggitt** [comment]
@anwu1219 in the end I decided it would be far better to add some proper unit tests rather than simply pushing through large networks. I've therefore added a zoo of ONNX networks each consisting of a single node (+ some constants), one for each layer type that we support.

There are two layers for which I've run into problems:
- A single constant layer runs into #637 
- A single sigmoid layer runs into a segfault in Marabou itself when trying to solve:
```
Running 1 testAddressSanitizer:DEADLYSIGNAL
=================================================================
==1511045==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x557671a36fd6 bp 0x7fff1c539f20 sp 0x7fff1c539ec0 T0)
==1511045==The signal is caused by a READ memory access.
==1511045==Hint: address points to the zero page.
    #0 0x557671a36fd6 in Tableau::storeState(TableauState&, TableauStateStorageLevel) const /home/matthew/Code/AISEC/Marabou/src/engine/Tableau.cpp:1632
    #1 0x5576718ff041 in Engine::storeState(EngineState&, TableauStateStorageLevel) const /home/matthew/Code/AISEC/Marabou/src/engine/Engine.cpp:1699
    #2 0x55767199a657 in PrecisionRestorer::storeInitialEngineState(IEngine const&) /home/matthew/Code/AISEC/Marabou/src/engine/PrecisionRestorer.cpp:27
    #3 0x5576718d8b45 in Engine::storeInitialEngineState() /home/matthew/Code/AISEC/Marabou/src/engine/Engine.cpp:2194
    #4 0x5576718fa592 in Engine::solve(unsigned int) /home/matthew/Code/AISEC/Marabou/src/engine/Engine.cpp:188
    #5 0x5576718cb017 in OnnxParserTestSuite::run_test(String, Vector<double>, Vector<double>) (/home/matthew/Code/AISEC/Marabou/build/tests/parser/Test_OnnxParser+0x54017)
    #6 0x5576718cce64 in OnnxParserTestSuite::test_sigmoid() (/home/matthew/Code/AISEC/Marabou/build/tests/parser/Test_OnnxParser+0x55e64)
    #7 0x5576718cd403 in TestDescription_OnnxParserTestSuite_test_sigmoid::runTest() (/home/matthew/Code/AISEC/Marabou/build/tests/parser/Test_OnnxParser+0x56403)
    #8 0x5576718b8fc8 in CxxTest::RealTestDescription::run() /home/matthew/Code/AISEC/Marabou/tools/cxxtest/cxxtest/RealDescriptions.cpp:96
    #9 0x5576718bfe7c in CxxTest::TestRunner::runSuite(CxxTest::SuiteDescription&) (/home/matthew/Code/AISEC/Marabou/build/tests/parser/Test_OnnxParser+0x48e7c)
    #10 0x5576718c048b in CxxTest::TestRunner::runWorld() (/home/matthew/Code/AISEC/Marabou/build/tests/parser/Test_OnnxParser+0x4948b)
    #11 0x5576718c0734 in CxxTest::TestRunner::runAllTests(CxxTest::TestListener&) (/home/matthew/Code/AISEC/Marabou/build/tests/parser/Test_OnnxParser+0x49734)
    #12 0x5576718ba161 in CxxTest::ErrorFormatter::run() /home/matthew/Code/AISEC/Marabou/tools/cxxtest/cxxtest/ErrorFormatter.h:47
    #13 0x5576718ba161 in main /home/matthew/Code/AISEC/Marabou/build/src/input_parsers/Test_OnnxParser.cc:16
    #14 0x7f5fcb029d8f in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #15 0x7f5fcb029e3f in __libc_start_main_impl ../csu/libc-start.c:392
    #16 0x5576718b6d74 in _start (/home/matthew/Code/AISEC/Marabou/build/tests/parser/Test_OnnxParser+0x3fd74)
```
Given the simplicity of the sigmoid parsing code, I don't really think it can be the fault of the ONNX parser as it's identical to that for the python ONNX parser sigmoid implementation. 

Given that sigmoid was already implemented already, it not working doesn't make things any worse than they previously were. I therefore propose that we file an issue for that, and this PR is ready to merge in, subject to your final approval?

**Comment by MatthewDaggitt** [comment]
Oh I forgot to say that I updated the version of the C++ ONNX library we build against, so that we use the same version of ONNX for both the Python and the C++ backends :+1:

**Comment by MatthewDaggitt** [comment]
Same non-deterministic test failure as always...

**Comment by anwu1219** [comment]
Hi @MatthewDaggitt , thanks a lot for the added test! The PR LGTM!

**Comment by anwu1219** [comment]
> @anwu1219 in the end I decided it would be far better to add some proper unit tests rather than simply pushing through large networks. I've therefore added a zoo of ONNX networks each consisting of a single node (+ some constants), one for each layer type that we support.
> 
> There are two layers for which I've run into problems:
> 
> * A single constant layer runs into [Minor bug in Python/C++ parsing of ONNX networks with constant outputs #637](https://github.com/NeuralNetworkVerification/Marabou/issues/637)
> * A single sigmoid layer runs into a segfault in Marabou itself when trying to solve:
> 
> ```
> Running 1 testAddressSanitizer:DEADLYSIGNAL
> =================================================================
> ==1511045==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x557671a36fd6 bp 0x7fff1c539f20 sp 0x7fff1c539ec0 T0)
> ==1511045==The signal is caused by a READ memory access.
> ==1511045==Hint: address points to the zero page.
>     #0 0x557671a36fd6 in Tableau::storeState(TableauState&, TableauStateStorageLevel) const /home/matthew/Code/AISEC/Marabou/src/engine/Tableau.cpp:1632
>     #1 0x5576718ff041 in Engine::storeState(EngineState&, TableauStateStorageLevel) const /home/matthew/Code/AISEC/Marabou/src/engine/Engine.cpp:1699
>     #2 0x55767199a657 in PrecisionRestorer::storeInitialEngineState(IEngine const&) /home/matthew/Code/AISEC/Marabou/src/engine/PrecisionRestorer.cpp:27
>     #3 0x5576718d8b45 in Engine::storeInitialEngineState() /home/matthew/Code/AISEC/Marabou/src/engine/Engine.cpp:2194
>     #4 0x5576718fa592 in Engine::solve(unsigned int) /home/matthew/Code/AISEC/Marabou/src/engine/Engine.cpp:188
>     #5 0x5576718cb017 in OnnxParserTestSuite::run_test(String, Vector<double>, Vector<double>) (/home/matthew/Code/AISEC/Marabou/build/tests/parser/Test_OnnxParser+0x54017)
>     #6 0x5576718cce64 in OnnxParserTestSuite::test_sigmoid() (/home/matthew/Code/AISEC/Marabou/build/tests/parser/Test_OnnxParser+0x55e64)
>     #7 0x5576718cd403 in TestDescription_OnnxParserTestSuite_test_sigmoid::runTest() (/home/matthew/Code/AISEC/Marabou/build/tests/parser/Test_OnnxParser+0x56403)
>     #8 0x5576718b8fc8 in CxxTest::RealTestDescription::run() /home/matthew/Code/AISEC/Marabou/tools/cxxtest/cxxtest/RealDescriptions.cpp:96
>     #9 0x5576718bfe7c in CxxTest::TestRunner::runSuite(CxxTest::SuiteDescription&) (/home/matthew/Code/AISEC/Marabou/build/tests/parser/Test_OnnxParser+0x48e7c)
>     #10 0x5576718c048b in CxxTest::TestRunner::runWorld() (/home/matthew/Code/AISEC/Marabou/build/tests/parser/Test_OnnxParser+0x4948b)
>     #11 0x5576718c0734 in CxxTest::TestRunner::runAllTests(CxxTest::TestListener&) (/home/matthew/Code/AISEC/Marabou/build/tests/parser/Test_OnnxParser+0x49734)
>     #12 0x5576718ba161 in CxxTest::ErrorFormatter::run() /home/matthew/Code/AISEC/Marabou/tools/cxxtest/cxxtest/ErrorFormatter.h:47
>     #13 0x5576718ba161 in main /home/matthew/Code/AISEC/Marabou/build/src/input_parsers/Test_OnnxParser.cc:16
>     #14 0x7f5fcb029d8f in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
>     #15 0x7f5fcb029e3f in __libc_start_main_impl ../csu/libc-start.c:392
>     #16 0x5576718b6d74 in _start (/home/matthew/Code/AISEC/Marabou/build/tests/parser/Test_OnnxParser+0x3fd74)
> ```
> 
> Given the simplicity of the sigmoid parsing code, I don't really think it can be the fault of the ONNX parser as it's identical to that for the python ONNX parser sigmoid implementation.
> 
> Given that sigmoid was already implemented already, it not working doesn't make things any worse than they previously were. I therefore propose that we file an issue for that, and this PR is ready to merge in, subject to your final approval?

Interesting.. Yes, please file an issue for this. Thank you!

**Review by anwu1219 — APPROVED** [review]

**Pull request merged**